### PR TITLE
[driver] "waitFor" command in place of broken "exists"

### PR DESCRIPTION
**Pull request by yjbanov**
- Replace the driver-side polling "waitFor" method with one executed on the extension side
- Delete the half-implemented "Exists" command (currently breaks the `sample_app` integration test)

### CODE REVIEW

**Comment by Hixie** [comment]
Rather than something based on timeouts, would it be possible to use positive feedback from the app saying "ok, we've pumped a frame"? A timeout in a test is a huge red flag IMHO.

**Comment by yjbanov** [comment]
@Hixie 

> would it be possible to use positive feedback from the app saying "ok, we've pumped a frame"?

Interesting. Do we reach into the framework on every frame, or only when something is dirty and needs rebuilding? Since Flutter Driver is intended for functional testing it needs to be resilient to application bugs. For example, imagine tapping on a button that does not result in any `setState`, therefore not triggering a rebuild, or perhaps it schedules a timer, so the state is expected to change after a delay. If no frame is pumped, we may miss the notification.

**Comment by Hixie** [comment]
Yeah that makes sense. We don't guarantee that you'll be called every frame. We could guarantee it by triggering a frame whenever we expect one, but that might change the behavior too much.

Anyway, using the callback is a good way to determine that something is worth checking again without having to poll, but I agree that you'll still want a timeout.

**Comment by yjbanov** [comment]
Make the finders work off of frames rather than a periodic timer.

PTAL: https://github.com/flutter/flutter/pull/3373/commits/370ff93786e30ea1d349165ff27c73d4cc7a9b31

File: packages/flutter_driver/lib/src/extension.dart [hunk 1]
```diff
@@ -112,16 +110,42 @@ class FlutterDriverExtension {
     }
   }
 
+  Stream<Duration> _onFrameReadyStream;
+  Stream<Duration> get _onFrameReady {
+    if (_onFrameReadyStream == null) {
+      // Lazy-initialize the frame callback because the renderer is not yet
+      // available at the time the extension is registered.
+      StreamController<Duration> frameReadyController = new StreamController<Duration>.broadcast(sync: true);
+      Renderer.instance.addPersistentFrameCallback((Duration timestamp) {
```
Comment (Hixie) [inline packages/flutter_driver/lib/src/extension.dart]: Scheduler, not Renderer

Comment (Hixie) [inline packages/flutter_driver/lib/src/extension.dart]: i mean it's the same object in principle, but the Scheduler is the one that actually implements it

Comment (yjbanov) [inline packages/flutter_driver/lib/src/extension.dart]: Done.

**Comment by Hixie** [comment]
LGTM.

FWIW, I find it confusing the way the same file includes code from both sides of the Web Socket. I guess that's inevitable since we want to be reusing so much code, but I would have expected an architecture where the flutter app server library and the flutter test client library both depended on a common set of libraries but otherwise were in very separate files. I think having that structure would help with moving the extensions into flutter so that we don't have to have a special driver app that turns on the extensions.

**Comment by yjbanov** [comment]
We could split it into multiple packages (`flutter_driver`, `flutter_driver_common`, the extension lives in `flutter`), or perhaps separate directories under the `flutter_driver` package, e.g.:

```
pkg: flutter_driver
lib/
  driver.dart
  extension.dart
  src/
    driver/
      ... test-side code ...
    extension/
      ... extension-side code ...
    common/
      ... code shared between driver and extension ...
```

@Hixie, do you have a preference?

**Comment by Hixie** [comment]
If we're going to make this code run in any app that is in profile mode, without explicit opt-in, then I think I'd put the common code (the code that needs to run every app) in `//flutter/packages/flutter/lib/src/instrumentation` or some such.

**Comment by Hixie** [comment]
or even just in `//flutter/packages/flutter/lib/src/widgets/instrumentation.dart`.

**Comment by Hixie** [comment]
(since the widgets binding.dart file presumably needs to know about it to register the extension)

**Pull request merged**